stdlib/time.mt

```diff
@@ -1,18 +1,177 @@
+        num100 = 3
+      end
+      total_days -= num100 * DAYS_PER_100_YEARS
+  
+      num4 = total_days / DAYS_PER_4_YEARS
+      total_days -= num4 * DAYS_PER_4_YEARS
+  
+      num_years = total_days / 365
+  
+      when num_years == 4 # leap
+        num_years = 3
+      end
+  
+      year = num400 * 400 + num100 * 100 + num4 * 4 + num_years + 1
+  
+      total_days -= num_years * 365
+      day_year = total_days + 1
+  
+      dec_31_leap = num100 == 3 || num4 != 24
+      when num_years == 3 && dec_31_leap # 31 dec leapyear
+        days = DAYS_MONTH_LEAP
+      end
+  
+      while total_days >= days[m]
+        total_days -= days[m]
+        m += 1
+      end
+  
+      month = m
+      day = total_days + 1
+  
+      [year, month, day, day_year]
+    end
+
+    def convert_to_seconds(year, month, day, hour, minute, second)
+      days = absolute_days(year, month, day)
+      seconds = 1 *
+                SECONDS_PER_DAY * days +
+                SECONDS_PER_HOUR * hour +
+                SECONDS_PER_MINUTE * minute +
+                second
+    end
+  end
+
+  def initialize(seconds, nanoseconds)
+    @seconds = seconds
+    @nanoseconds = nanoseconds
+  end
+
+  def initialize(year, month, day)
+    @seconds = Util.convert_to_seconds(year, month, day, 0, 0, 0)
+    @nanoseconds = 0
+  end
+
+  def initialize(year, month, day, hour)
+    @seconds = Util.convert_to_seconds(year, month, day, hour, 0, 0)
+    @nanoseconds = 0
+  end
+
+  def initialize(year, month, day, hour, minute)
+    @seconds = Util.convert_to_seconds(year, month, day, hour, minute, 0)
+    @nanoseconds = 0
+  end
+
   def initialize(year, month, day, hour, minute, second)
-    @year = year
-    @month = month
-    @day = day
-    @hour = hour
-    @minute = minute
-    @second = second
+    @seconds = Util.convert_to_seconds(year, month, day, hour, minute, second)
+    @nanoseconds = 0
+  end
+
+  def initialize(year, month, day, hour, minute, second, nanosecond)
+    @seconds = Util.convert_to_seconds(year, month, day, hour, minute, second)
+    @nanoseconds = nanosecond
   end
 
   # Only getters since a given Time is immutable
-  def year; @year; end
-  def month; @month; end
-  def day; @day; end
-  def hour; @hour; end
-  def minute; @minute; end
-  def second; @second; end
+  def seconds; @seconds; end
+  def nanoseconds; @nanoseconds; end
+
+  def year
+    Util.year_month_day_day_year(@seconds)[0]
+  end
+
+  def month
+    Util.year_month_day_day_year(@seconds)[1]
+  end
+
+  def day
+    Util.year_month_day_day_year(@seconds)[2]
+  end
+
+  def hour
+    (@seconds % Util.SECONDS_PER_DAY) / Util.SECONDS_PER_HOUR
+  end
+
+  def minute
+    (@seconds % Util.SECONDS_PER_HOUR) / Util.SECONDS_PER_MINUTE
+  end
+
+  def second
+    @seconds % Util.SECONDS_PER_MINUTE
+  end
+
+  def millisecond
+    @nanoseconds / Util.NANOSECONDS_PER_MILLISECOND
+  end
+
+  def nanosecond
+    @nanoseconds
+  end
+
+  def -(other)
+    when other.type.to_s != "Time"
```

Comment:
This could be written cleaner as `def -(other : Time)` and then removing the conditional.

Right now, that would actually change the semantics of the code, but once #98 gets merged, it should be almost identical. Even then, I think I'd prefer the `No clause matches` error just so it doesn't get forgotten about in the future.

Reply:
Yeah that makes sense, my only issue is the No clause error is a Crystal error.  It is preventing the specs from running as is, I could comment out the spec for now, or have a second method which catches all other and raises

Reply:
I can get on board with the second one. Maybe just add a comment that the other clause is temporary until #98 gets resolved?